scripts/externalTests/test_helpers.py

```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env python3
+
+# ------------------------------------------------------------------------------
+# This file is part of solidity.
+#
+# solidity is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# solidity is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2023 solidity contributors.
+# ------------------------------------------------------------------------------
+
+import os
+import re
+import subprocess
+import sys
+from abc import abstractmethod, ABCMeta
+from argparse import ArgumentParser
+from dataclasses import dataclass, field
+from enum import Enum
+from pathlib import Path
+from shutil import rmtree
+from tempfile import mkdtemp
+from textwrap import dedent
+from typing import List, Set
+
+# Our scripts/ is not a proper Python package so we need to modify PYTHONPATH to import from it
+# pragma pylint: disable=import-error,wrong-import-position
+PROJECT_ROOT = Path(__file__).parents[2]
+sys.path.insert(0, f"{PROJECT_ROOT}/scripts/common")
+
+from git_helpers import git_commit_hash
+
+SOLC_FULL_VERSION_REGEX = re.compile(r"^[a-zA-Z: ]*(.*)$")
+SOLC_SHORT_VERSION_REGEX = re.compile(r"^([0-9.]+).*\+|\-$")
+
+evm_version = os.environ.get("DEFAULT_EVM")
+CURRENT_EVM_VERSION: str = evm_version if evm_version is not None else "shanghai"
+
+
+class SettingsPreset(Enum):
+    LEGACY_NO_OPTIMIZE = 'legacy-no-optimize'
+    IR_NO_OPTIMIZE = 'ir-no-optimize'
+    LEGACY_OPTIMIZE_EVM_ONLY = 'legacy-optimize-evm-only'
+    IR_OPTIMIZE_EVM_ONLY = 'ir-optimize-evm-only'
+    LEGACY_OPTIMIZE_EVM_YUL = 'legacy-optimize-evm+yul'
+    IR_OPTIMIZE_EVM_YUL = 'ir-optimize-evm+yul'
+
+
+@dataclass
+class TestConfig:
+    name: str
+    repo_url: str
+    ref_type: str
+    ref: str
+    build_dependency: str = field(default="nodejs")
+    compile_only_presets: List[SettingsPreset] = field(default_factory=list)
+    settings_presets: List[SettingsPreset] = field(default_factory=lambda: list(SettingsPreset))
+    evm_version: str = field(default=CURRENT_EVM_VERSION)
+
+    def selected_presets(self) -> Set[SettingsPreset]:
+        return set(self.compile_only_presets + self.settings_presets)
+
+
+class TestRunner(metaclass=ABCMeta):
```

Comment:
This isn't really a helper. Thinking more about the structure, I'd put `TestRunner` into a separate module, along with `run_test()`. Maybe also rename it to `BaseRunner`, `AbstractRunner` or `GenericTestRunner`. This is the backbone of the whole mechanism and having it separately would make it easier to grasp the overall control flow. Then I'd leave only actual helpers here.

Reply:
I moved it to the runners module.